bfvmm/include/hve/arch/intel_x64/exit_handler.h

```diff
@@ -133,66 +145,19 @@ class EXPORT_HVE exit_handler
     ///
     /// @param d The delegate being registered
     ///
-    VIRTUAL void add_exit_handler(
+    void add_exit_handler(
         const handler_delegate_t &d
     );
 
-    /// Handle
-    ///
-    /// Handles a VM exit. This function should only be called by the exit
-    /// handler entry function, which gets called when a VM exit occurs, and
-    /// then calls this function to handler the VM exit to the proper
-    /// handler.
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @param vcpu The vcpu this exit occured on
-    ///
-    /// @return True if the vmexit was successfully handled, false otherwise
-    bool handle(vcpu_t vcpu);
-
-private:
-
-    // These members provide the legacy exit handler mechanism, and will be
-    // deprecated slowly
-    std::list<init_handler_delegate_t> m_exit_handlers;
-    std::list<init_handler_delegate_t> m_init_handlers;
-    std::list<fini_handler_delegate_t> m_fini_handlers;
-    std::array<std::list<handler_delegate_t>, 64> m_exit_handlers_array;
-
-    // These members will provide the new exit handler mechanism
-    std::array<handler_delegate_t, 64> m_exit_delegates;
-    std::unique_ptr<cpuid::delegator> m_cpuid_delegator;
-    std::unique_ptr<nmi::delegator> m_nmi_delegator;
-    std::unique_ptr<cr::delegator> m_cr_delegator;
-    std::unique_ptr<invd::delegator> m_invd_delegator;
-    std::unique_ptr<msr::delegator> m_msr_delegator;
-    std::unique_ptr<fallback::delegator> m_fallback_delegator;
-
 public:
 
-    /// @cond
-
-    cpuid::delegator *cpuid_delegator() const
-    { return m_cpuid_delegator.get(); }
-
-    nmi::delegator *nmi_delegator() const
-    { return m_nmi_delegator.get(); }
+    static void handle(
```

Comment:
Why does this take an exit_handler as an argument instead of just using ```this```?

Reply:
Because it is a static function so there is no this. Technically RDI stores the "this" point if we wanted to use a member function and remove "static", but its really hard to read code like that as you really have to know a lot about the System V calling convention to understand how that works.

Reply:
I didn't even notice that this was a static function, thats even more confusing to me.

If there is going to be a single instance of an entry point for all vmexits, I would think that separating it completely from this class would be a better solution.

Reply:
I am not opposed to that. I would take that a step further and make it a C extern so that the asm code that references this function doesn't have any mangling in it. I will upload that change.

Reply:
Ok, so I remember why we did this. It was a friend function of an extern "C" function issue (and me simply not wanting to figure out the black magic to make that work). 

Fixed

Reply:
Nice! I really like this solution, because it gets rid of the name-mangling in the exit handler entry assembly. 👍